addons/calendar/models/calendar_recurrence.py

```diff
@@ -240,6 +241,47 @@ def _apply_recurrence(self, specific_values_creation=None, no_send_edit=False, g
         self.env['calendar.event'].with_context(no_mail_to_attendees=True, mail_create_nolog=True).create(event_vals)
         return detached_events
 
+    def _setup_alarms(self, recurrence_update=False):
+        """ Schedule cron triggers for future events
+        Create one ir.cron.trigger per recurrence.
+        :param recurrence_update: boolean: if true, update all recurrences in self, else only the recurrences
+               without trigger
+        """
+        now = fields.Datetime.now()
+        # get next events
+        self.env['calendar.event'].flush_model(fnames=['recurrence_id', 'start'])
+        if recurrence_update:
+            recurrence = self
+        else:
+            recurrence = self.filtered(lambda rec: not rec.trigger_id)
+        if not recurrence.calendar_event_ids.ids:
+            return
+        self.env.cr.execute("""
+            SELECT res.event_id,
+                   res.recurrence_id
+              FROM (
+                  SELECT id event_id,
+                         recurrence_id,
+                         RANK() OVER (PARTITION BY recurrence_id ORDER BY start ASC),
+                         start
+                    FROM calendar_event 
+                   WHERE start > %s 
+                     AND id IN %s
+                ORDER BY start ) AS res 
+                WHERE RANK = 1
+        """, (now, tuple(recurrence.calendar_event_ids.ids)))
```

Comment:
```sql
                  SELECT DISTINCT ON (recurrence_id) event_id, recurrence_id,
                    FROM calendar_event 
                   WHERE start > %s 
                     AND id IN %s
                ORDER BY start ASC
```
Should be enough (and there is a ',' missing after id in your query if you want to keep it)